File: apps/frontend/types/source-inventory.ts
export type SourceContentItemStatus =
  | "pending"
  | "mapped"
  | "reviewed"
  | "approved"
  | "rejected"
  | "generated"
  | "validated"
  | "possible_duplicate"
  | "fragmented"
  | "requires_review";

export type SourceContentItemImportance = "essential" | "relevant" | "complementary";

export type SourceContentItem = {
  id: string;
  project_id: string;
  project_file_id: string;
  item_code: string;
  title: string;
  source_text: string;
  normalized_content: string | null;
  content_type: string;
  page_start: number | null;
  page_end: number | null;
  source_order: number;
  importance: SourceContentItemImportance;
  status: SourceContentItemStatus;
  metadata_json: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export type SourceContentItemListResponse = {
  items: SourceContentItem[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
};

export type SourceContentItemBlock = {
  id: string;
  source_item_id: string;
  block_id: string;
  block_code: string;
  page_number: number;
  source_order: number;
  is_primary: boolean;
  created_at: string;
};

export type SourceItemDependency = {
  id: string;
  source_item_id: string;
  depends_on_source_item_id: string;
  dependency_type: string;
  created_at: string;
};

export type SourceContentItemDetail = SourceContentItem & {
  blocks: SourceContentItemBlock[];
  dependencies: SourceItemDependency[];
  dependents: SourceItemDependency[];
};

export type SourceInventorySummary = {
  project_id: string;
  project_file_id: string;
  status: string;
  inventory_version: number;
  total_pages: number;
  pages_processed: number;
  pages_not_processed: number;
  pages_requires_ocr: number;
  total_blocks: number;
  blocks_analyzed: number;
  blocks_ignored: number;
  total_chunks: number;
  chunks_completed: number;
  chunks_failed: number;
  total_items: number;
  items_by_type: Record<string, number>;
  items_by_importance: Record<string, number>;
  possible_duplicates: number;
  fragmented_items: number;
  requires_review_items: number;
  approved_items: number;
  rejected_items: number;
  page_coverage_percentage: number;
  block_coverage_percentage: number;
  model_name: string | null;
  prompt_version: string | null;
  generated_at: string | null;
};

export type SourceInventoryReprocessMode =
  | "generate_if_missing"
  | "reprocess_failed"
  | "reprocess_pages"
  | "full_rebuild"
  | "validate_only";

export type SourceInventoryItemManualUpdate = {
  title?: string;
  normalized_content?: string;
  content_type?: string;
  importance?: string;
  review_note?: string;
};

export type InventoryValidationIssue = {
  source_item_id: string | null;
  item_code: string | null;
  issue_type: string;
  message: string;
};

export type InventoryValidationResult = {
  status: string;
  total_items: number;
  valid_items: number;
  invalid_items: number;
  issues: InventoryValidationIssue[];
};

File: apps/backend/app/schemas/source_inventory.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from app.schemas.source_content_item import (
    SOURCE_CONTENT_ITEM_CONTENT_TYPES,
    SOURCE_CONTENT_ITEM_IMPORTANCE_LEVELS,
    SourceContentItemResponse,
)

REPROCESS_MODES = {
    "generate_if_missing",
    "reprocess_failed",
    "reprocess_pages",
    "full_rebuild",
    "validate_only",
}

DEPENDENCY_TYPES = {
    "explains",
    "exemplifies",
    "continues",
    "qualifies",
    "contradicts",
    "depends_on",
    "exception_to",
    "part_of",
}


class SourceInventoryGenerateRequest(BaseModel):
    force: bool = False
    continue_with_alerts: bool = False


class SourceInventoryReprocessRequest(BaseModel):
    mode: str = "reprocess_failed"
    page_numbers: list[int] | None = None
    continue_with_alerts: bool = False

    @model_validator(mode="after")
    def validate_mode(self) -> "SourceInventoryReprocessRequest":
        if self.mode not in REPROCESS_MODES:
            raise ValueError(f"mode invalido: {self.mode}")
        if self.mode == "reprocess_pages" and not self.page_numbers:
            raise ValueError("page_numbers e obrigatorio quando mode='reprocess_pages'.")
        if self.page_numbers is not None and any(p <= 0 for p in self.page_numbers):
            raise ValueError("page_numbers deve conter apenas valores maiores que zero.")
        return self


class SourceInventoryItemManualUpdate(BaseModel):
    """Edicao humana de um item do inventario. source_text original nao e editavel aqui
    de proposito (rastreabilidade com o documento fonte deve ser preservada)."""

    title: str | None = Field(default=None, min_length=1, max_length=500)
    normalized_content: str | None = Field(default=None, min_length=1)
    content_type: str | None = None
    importance: str | None = None
    review_note: str | None = None

    @model_validator(mode="after")
    def validate_enums(self) -> "SourceInventoryItemManualUpdate":
        if self.content_type is not None and self.content_type not in SOURCE_CONTENT_ITEM_CONTENT_TYPES:
            raise ValueError(f"content_type invalido: {self.content_type}")
        if self.importance is not None and self.importance not in SOURCE_CONTENT_ITEM_IMPORTANCE_LEVELS:
            raise ValueError(f"importance invalido: {self.importance}")
        return self


class SourceContentItemBlockResponse(BaseModel):
    id: UUID
    source_item_id: UUID
    block_id: UUID
    block_code: str
    page_number: int
    source_order: int
    is_primary: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class SourceItemDependencyResponse(BaseModel):
    id: UUID
    source_item_id: UUID
    depends_on_source_item_id: UUID
    dependency_type: str
    created_at: datetime

    model_config = {"from_attributes": True}


class SourceContentItemDetailResponse(SourceContentItemResponse):
    blocks: list[SourceContentItemBlockResponse] = []
    dependencies: list[SourceItemDependencyResponse] = []
    dependents: list[SourceItemDependencyResponse] = []


class SourceInventorySummary(BaseModel):
    project_id: UUID
    project_file_id: UUID
    status: str = "not_started"
    inventory_version: int = 0
    total_pages: int = 0
    pages_processed: int = 0
    pages_not_processed: int = 0
    pages_requires_ocr: int = 0
    total_blocks: int = 0
    blocks_analyzed: int = 0
    blocks_ignored: int = 0
    total_chunks: int = 0
    chunks_completed: int = 0
    chunks_failed: int = 0
    total_items: int = 0
    items_by_type: dict[str, int] = {}
    items_by_importance: dict[str, int] = {}
    possible_duplicates: int = 0
    fragmented_items: int = 0
    requires_review_items: int = 0
    approved_items: int = 0
    rejected_items: int = 0
    page_coverage_percentage: float = 0.0
    block_coverage_percentage: float = 0.0
    model_name: str | None = None
    prompt_version: str | None = None
    generated_at: datetime | None = None


class InventoryValidationIssue(BaseModel):
    source_item_id: UUID | None = None
    item_code: str | None = None
    issue_type: str
    message: str


class InventoryValidationResult(BaseModel):
    status: str
    total_items: int
    valid_items: int
    invalid_items: int
    issues: list[InventoryValidationIssue] = []
